snuba/datasets/entity.py

```diff
@@ -156,3 +175,106 @@ def describe(self) -> Description:
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def to_dict(self) -> Mapping[str, Any]:
+        raise NotImplementedError
+
+    @abstractmethod
+    def set_org(self, data_dict: Mapping[str, Any] = {}) -> EntitySubscription:
+        raise NotImplementedError
+
+
+class EntitySubscriptionValidation(EntitySubscription):
+    def __init__(
+        self,
+        max_allowed_aggregations: int = 1,
+        disallowed_aggregations: Sequence[str] = ["groupby", "having", "orderby"],
+    ) -> None:
+        self.max_allowed_aggregations: int = max_allowed_aggregations
+        self.disallowed_aggregations: Sequence[str] = disallowed_aggregations
+
+    def validate_query(self, query: Union[CompositeQuery[EntityDS], Query]) -> None:
+        # Import get_entity() when used, not at import time to avoid circular imports
+        from snuba.datasets.entities.factory import get_entity
+
+        # TODO: Support composite queries with multiple entities.
+        from_clause = query.get_from_clause()
+        if not isinstance(from_clause, EntityDS):
+            raise InvalidSubscriptionError("Only simple queries are supported")
+        entity = get_entity(from_clause.key)
+
+        SubscriptionAllowedClausesValidator(
+            self.max_allowed_aggregations, self.disallowed_aggregations
+        ).validate(query)
+        if entity.required_time_column:
+            NoTimeBasedConditionValidator(entity.required_time_column).validate(query)
+
+
+class BaseEntitySubscription(EntitySubscriptionValidation):
+    def __init__(
+        self,
+        max_allowed_aggregations: Optional[int] = None,
+        disallowed_aggregations: Optional[Sequence[str]] = None,
+        data_dict: Mapping[str, Any] = {},
+    ) -> None:
+        super().__init__()
+        if max_allowed_aggregations and disallowed_aggregations:
+            super().__init__(max_allowed_aggregations, disallowed_aggregations)
+        self.set_org(data_dict)
+
+    def set_org(self, data_dict: Mapping[str, Any] = {}) -> BaseEntitySubscription:
+        if "organization" in data_dict:
+            try:
+                self.organization: int = data_dict["organization"]
+            except KeyError:
+                raise InvalidQueryException(
+                    "organization param is required for any query over sessions entity"
+                )
+        return self
+
+    def get_entity_subscription_conditions_for_snql(
+        self, offset: Optional[int] = None
+    ) -> Sequence[Expression]:
+        if self.organization:
+            return [
+                binary_condition(
+                    ConditionFunctions.EQ,
+                    Column(None, None, "org_id"),
+                    Literal(None, self.organization),
+                ),
+            ]
+        return []
+
+    def to_dict(self) -> Mapping[str, Any]:
+        if self.organization:
+            return {"organization": self.organization}
+        return {}
```

Comment:
Why does this need to be 3 classes when the inheritance tree is just a straight line?

Reply:
Also `Base*` by convention implies a base class that is to be extended. Is this one of those cases or is it just an `EntitySubscription`?

Reply:
Updated back to the original structure. The idea behind the base class is so that we can extend the functionality of subscriptions the same way we extend things like translation mappers. Whether or not that should extend from `EntitySubscription` or the newly created `BaseEntitySubscription` is still to be determined.